sklearn/ensemble/_hist_gradient_boosting/_histogram_pool.py

```diff
@@ -0,0 +1,35 @@
+import numpy as np
+from .common import HISTOGRAM_DTYPE
+
+
+class HistogramsPool:
+    """Histograms pool to be used by the growers. The pool allocates and
+    returns references to the histograms to the caller. When the `reset` method
+    is called all, the allocated histograms will be avaliable for the next
+    grower to use. New histograms will be allocated when there are no
+    histograms left in the avaliable_pool. HistogramsPool is used for memory
+    allocation/management and that no building logic is done. The building will
+    be done in HistogramBuilder.
+    """
+    def __init__(self, n_features, n_bins):
+        self.n_features = n_features
+        self.n_bins = n_bins
+        self.avaliable_pool = []
+        self.used_pool = []
+
+    def reset(self):
+        """Reset the pool."""
+        # Only set histograms that were used to zero
+        self.avaliable_pool.extend(self.used_pool)
+        self.used_pool = []
+
+    def get(self):
+        """Return a histograms object used by grower."""
+        if self.avaliable_pool:
+            histograms = self.avaliable_pool.pop()
```

Comment:
can you confirm that `pop`ing a list will not try to downsize the list at some point?

Reply:
When a list is resize the copies are not deep:

```py
import numpy as np

first_elm = np.ones(1)
items = [first_elm]
for _ in range(100):
    items.append(np.empty(1))

assert first_elm is items[0]

for _ in range(100):
    items.pop()
    
assert first_elm is items[0]
```